java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -11,15 +11,42 @@ private import FlowSummaryImpl as FlowSummaryImpl
 private import DataFlowImplConsistency
 import DataFlowNodes::Private
 
-private newtype TReturnKind = TNormalReturnKind()
+private newtype TReturnKind =
+  TNormalReturnKind() or
+  TJumpReturnKind(Call target)
```

Comment:
I think this ought to be `Callable` instead of `Call`.

Reply:
That's what C# does as well, but I believe Tony needed finer granularity.

Reply:
Yes, the whole need for jump summaries comes from our intention to support the following scenario:

```java
class Activity1 {
  public void onCreate() {
    Intent intent = new Intent(this, Activity2.class);
    intent.putExtra("key", source());
    startActivities(new Intent[]{intent});
  }
}

class Activity2 {
  public void onCreate() {
    sink(getIntent().getExtra("key")); // vulnerable
  }
}

class Activity3 {
  public void onCreate() {
    sink(getIntent().getExtra("key")); // safe
  }
}
```

We need to add an edge to the specific `getIntent` calls in the activities targeted by the appropriate `Intent` constructors only — if we used `Callable` instead, all `getIntent` calls would have an edge regardless of whether they were targeted or not.

This is usually the kind of step that isn't covered well by MaD and needs regular CodeQL, but since in this case we need precise container flow (because of the `putExtra`/`getExtra` shenanigans), we can't use a normal `AdditionalTaintStep` (and neither an `AdditionalValueStep`, because of the type change from `Intent[]` to `Intent`).

Reply:
Hmm, but then this PR is not even precise enough, right? Because you still only have one call target for the entry points (i.e. the calls to `startActivities`). So if there's another call like the following then you'll get FP flow.
```
  public void onCreate() {
    Intent intent = new Intent(this, Activity3.class); // note Activity3 here
    intent.putExtra("key", safeValue); // no source here
    startActivities(new Intent[]{intent}); // this might induce a jump-return to Activity3, right
  }
```
I think we need to take a step back and reconsider how to best solve this.

Reply:
How would you currently envision identifying the `Call` in `Activity2`, but not the one in `Activity3`? I'm guessing some kind of local flow constraint from an `Intent` to `startActivities`? But then you already have hold of the `intent` variable, so it seems like we're fairly close to just being able to use `AdditionalValueStep`.

Reply:
Yes, that's the current approach, local data flow from the `Intent` constructor (where you have the target activity information) to the `startActivities` argument. Would you suggest creating the step from one of the accesses to that `intent` before it's added to the array that would end up as the `startActivities` argument, instead of from the argument itself? I guess it would work, but the result path would look confusing.

Reply:
And yes, your consideration about FPs is correct, it's a shortcoming we discussed with @hvitved and concluded that for that to work 100% correctly we need both edges of the summary to be calls instead of callables, which is something not currently supported.